polkadot/runtime/parachains/src/paras_inherent/mod.rs

```diff
@@ -997,54 +1005,58 @@ fn sanitize_backed_candidate_v2<T: crate::inclusion::Config>(
 		return false
 	}
 
-	let Some(session_index) = candidate.descriptor().session_index() else {
+	// Get the claim queue snapshot at the candidate relay parent.
+	let Some((rp_info, _)) =
+		allowed_relay_parents.acquire_info(candidate.descriptor().relay_parent(), None)
+	else {
 		log::debug!(
 			target: LOG_TARGET,
-			"Invalid V2 candidate receipt {:?} for paraid {:?}, missing session index.",
+			"Relay parent {:?} for candidate {:?} is not in the allowed relay parents.",
+			candidate.descriptor().relay_parent(),
 			candidate.candidate().hash(),
-			candidate.descriptor().para_id(),
 		);
 		return false
 	};
 
-	// Check if session index is equal to current session index.
-	if session_index != shared::CurrentSessionIndex::<T>::get() {
+	if let Err(err) = candidate.candidate().check_ump_signals(&rp_info.claim_queue) {
 		log::debug!(
 			target: LOG_TARGET,
-			"Dropping V2 candidate receipt {:?} for paraid {:?}, invalid session index {}, current session {}",
+			"UMP signal check failed: {:?}. Dropping candidate {:?} for paraid {:?}.",
+			err,
 			candidate.candidate().hash(),
-			candidate.descriptor().para_id(),
-			session_index,
-			shared::CurrentSessionIndex::<T>::get()
+			candidate.descriptor().para_id()
 		);
 		return false
 	}
 
-	// Get the claim queue snapshot at the candidate relay parent.
-	let Some((rp_info, _)) =
-		allowed_relay_parents.acquire_info(candidate.descriptor().relay_parent(), None)
-	else {
+	if descriptor_version == CandidateDescriptorVersion::V1 {
+		// Nothing more to check for v1 descriptors.
+		return true
+	}
+
+	let Some(session_index) = candidate.descriptor().session_index() else {
 		log::debug!(
 			target: LOG_TARGET,
-			"Relay parent {:?} for candidate {:?} is not in the allowed relay parents.",
-			candidate.descriptor().relay_parent(),
+			"Invalid V2 candidate receipt {:?} for paraid {:?}, missing session index.",
 			candidate.candidate().hash(),
+			candidate.descriptor().para_id(),
 		);
 		return false
 	};
 
-	// Check validity of `core_index`.
-	if let Err(err) = candidate.candidate().check_core_index(&rp_info.claim_queue) {
+	// Check if session index is equal to current session index.
+	if session_index != shared::CurrentSessionIndex::<T>::get() {
```

Comment:
We should avoid having storage of another pallet accessible directly, but instead expose any needed data via functions. 

Case in point, we even have a `set_session_index` available only for [test](https://github.com/paritytech/polkadot-sdk/blob/b38c15868b1e00f63c5b92f3e210f6e7af1eae6b/polkadot/runtime/parachains/src/shared.rs#L264), which is kind of useless if the storage is publicly accessible anyways.

Reply:
I agree but this is quite unrelated. We have many ideas for refactoring the relay chain runtime modules. Maybe we should create a big issue with all the ideas, this being one of them: No public storage items